### PR TITLE
Fix Route Locator for both ZuulServlet and DispatcherServlet scenarios

### PR DESCRIPTION
**Pull request by aivans**
Both execution realms must work.

I created a ServletDetectionFilter for early detection for the servlet realm.
Route Locator will use this information for finding the route at runtime.

### CODE REVIEW

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulProxyConfiguration.java [hunk 1]
```diff
@@ -91,6 +92,11 @@ public DiscoveryClientRouteLocator routeLocator() {
 	}
 
 	// pre filters
+    @Bean
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/ZuulProxyConfiguration.java]: Formatting: we use tabs, not spaces.

**Comment by spencergibb** [comment]
Can you explain more of the problem you are trying to fix?

**Comment by aivans** [comment]
At the moment Zuul can function in 2 realms:
- DispatcherServlet
- ZuulServlet
Since this commit https://github.com/spring-cloud/spring-cloud-netflix/commit/053296de939831c4c72f6871d779f7cf4122c2c7 everything works nicely with jsut sending requests to /zuul, especially for big files. Basically it behaves as Netflix Zuul with the nice features brought by Spring (dependency injction, etc), plus the nice Admin actuator endpoints.

In my attempt to route eveything to ZuulServlet, I played with ```server.servletPath``` and ```zuul.servletPath``` and I noticed that ```simpleRouteLocator``` doesn't work properly. Myself I did a pull request on that a while ago to fix other issues. Basicaly we need to differentiate at runtime in a lot of places that the request goes through ZuulServlet directly or through DispatcherServlet. SimpleRouteLocator is one such case. We need to know at runtime the ```servletPath`` to strip from the path in order to match the route. the DispatcherServlet servletPath is provided through constructor, zuulProperties the same and I added a way to detect the servlet .

**Comment by aivans** [comment]
@spencergibb So, the short explanation would be (after a good sleep): routing through ZuulServlet directly will not work if the DispatcherServlet servletPath is changed  (will differ from "/").
Long story: we need a way to differentiate end to end from ZuulServlet execution and Dispatcherservlet execution.

**Comment by aivans** [comment]
Hi guys,

Can you please take a look at this and see whether something can be improved?

Thanks

**Comment by dsyer** [comment]
Did you fill in the contributor's agreement (link in README)?

**Comment by aivans** [comment]
yes, a  few months ago

**Comment by dsyer** [comment]
I see test failures (despite the travis checks being green). Can you rebase onto master and confirm that it works for you?

**Comment by aivans** [comment]
Looking into it

**Comment by aivans** [comment]
Good to go now.
The RequestContext as dirty between tests.

**Comment by dsyer** [comment]
Different failures now. It might be unrelated to your change, but if you can see how to fix it I can merge:

```
Results :

Failed tests: 
  ProxyRequestHelperTests.debug:78 
Expected: "{}"
     but: was null

Tests run: 252, Failures: 1, Errors: 0, Skipped: 2
```

**Comment by aivans** [comment]
weird, will do.

Works locally.

**Comment by aivans** [comment]
I am a bit puzzled because I see in the Travis console this:
Tests run: 249, Failures: 0, Errors: 0, Skipped: 2

Vs 252 which you reported and I can see locally.

The difference comes from here:
Running org.springframework.cloud.netflix.zuul.filters.discovery.DiscoveryClientRouteLocatorTests

Tests run: 36, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.142 sec - in org.springframework.cloud.netflix.zuul.filters.discovery.DiscoveryClientRouteLocatorTests

There are 39 tests in that class, I added 3 today.

**Pull request closed without merging**